relentless/simulate/lammps.py

```diff
@@ -0,0 +1,870 @@
+import abc
+import os
+
+import numpy as np
+
+from relentless.core.ensemble import RDF
+from relentless.core.volume import TriclinicBox
+from . import simulate
+
+try:
+    import lammps
+    _lammps_found = True
+except ImportError:
+    _lammps_found = False
+
+class LAMMPS(simulate.Simulation):
+    """:py:class:`Simulation` using LAMMPS framework.
+
+    Raises
+    ------
+        If the `lammps` package is not found, or is not version 29 Oct 2020 or newer.
+
+    """
+    def __init__(self, operations, quiet=True, **options):
+        if not _lammps_found:
+            raise ImportError('LAMMPS not found.')
+
+        super().__init__(operations,**options)
+        self.quiet = quiet
+
+    def _new_instance(self, ensemble, potentials, directory):
+        sim = super()._new_instance(ensemble,potentials,directory)
+
+        if self.quiet:
+            # create lammps instance with all output disabled
+            launch_args = ['-echo','none',
+                           '-log','none',
+                           '-screen','none',
+                           '-nocite']
+        else:
+            launch_args = ['-echo','screen',
+                           '-log', sim.directory.file('log.lammps'),
+                           '-nocite']
+
+        sim.lammps = lammps.lammps(cmdargs=launch_args)
+        if sim.lammps.version() < 20201029:
+            raise ImportError('Only LAMMPS 29 Oct 2020 or newer is supported.')
+
+        # lammps uses 1-indexed ints for types, so build mapping in both direction
+        sim.type_map = {}
+        sim.typeid_map = {}
+        for i,t in enumerate(sim.ensemble.types):
+            sim.type_map[t] = i+1
+            sim.typeid_map[sim.type_map[t]] = t
+
+        return sim
+
+class LAMMPSOperation(simulate.SimulationOperation):
+    _fix_counter = 1
+
+    def __call__(self, sim):
+        """Evaluates the LAMMPS simulation operation.
+
+        Parameters
+        ----------
+        sim : :py:class:`Simulation`
+            The simulation object.
+
+        """
+        cmds = self.to_commands(sim)
+        sim.lammps.commands_list(cmds)
+
+    @classmethod
+    def new_fix_id(cls):
+        # sets fix_id for LAMMPS fixes
+        idx = int(cls._fix_counter)
+        cls._fix_counter += 1
+        return idx
+
+    @abc.abstractmethod
+    def to_commands(self, sim):
```

Comment:
Document that this is the method that deriving classes should override, and it should return a list of valid LAMMPS commands.